Feature/ Photo upload

mattchx opened the pull request:
**Fully functional Photo Upload**
- image is previewed before upload
- saving changes posts the image to S3 bucket which returns a public url that is used in Avatar as `src`
- React Context is used to populate the image in the dashboard
- Material UI Dialog is used for upload modal

mattchx committed:
installed dependencies

mattchx committed:
Create ImageUploader.js

mattchx committed:
Update NavBar.js

mattchx committed:
BE: Added API and route to App.js

mattchx committed:
started image preview function

mattchx committed:
Image preview working

- uses useEffect
* upload to bucket returning 400 error

mattchx committed:
fixed upload error 400

mattchx committed:
AWS filename is sent from BE to FE

- the file name and bucket can be used to fetch the profile photo of each user

mattchx committed:
upload display photo function is live

- uses react context

mattchx committed:
code clean up

mattchx committed:
simplified context

WhoaNellie commented on client/src/context/UserContext.js:
```diff
@@ -0,0 +1,23 @@
+export const UserContext = createContext();
+
+export function UserContextProvider(props) {
+  const [url, setUrl] = useState('');
```
Might consider renaming this var to sometime like imgUrl/setImgUrl after we merge because there are some vars named url that deal with the users' unique urls. Depends on if they'll end up colliding on any pages though I suppose.

mattchx replied:
Sure that's an easy fix.

WhoaNellie commented on client/src/components/SelectTabs/EventTypesTab.js:
```diff
@@ -267,7 +265,7 @@ export default function EventTypesTab() {
             <Avatar
               className={classes.avatar}
-              src={ProfileImage}
+              src=''
```
Is this meant to be src={url} or does this work somehow?

mattchx replied:
React Context would not pass the updated url to `EventTypesTab.js` so I removed the var until we can get a fix.

dproc96 reviewed (changes requested):
It looks like you haven't saved the image to the actual profile document, so it doesn't persist. Please see Nellie's comments as well. Also take a look at the console, I noticed quite a few errors when I ran the app so if those are related to this feature please address them before pushing.

mattchx committed:
update variables names related to UserContext : photoUrl

*also added plus icon to add event type

mattchx committed:
photoUrl in User Schema recieves AWS url from BE

mattchx committed:
Changed name of UserContext.js to Context.js

mattchx commented:
I've addressed the concerns outlined by my reviewers.
Please take a look at my changes. Pay particular attention to my `db.User.updateOne()` function in the BE route.

* I have not corrected the FE compile errors as those are coming from other components unrelated to mine. I will merge dev into my branch when the latest version is approved and merged. I believe this will correct most/all of those errors. 

Thank you!

alvyjudy commented on server/routes/imageUploaderAPI.js:
```diff
@@ -0,0 +1,58 @@
+      if (req.file == undefined) {
+        res.status(400).send('There was a error with your upload paramaters');
+      } else {
+        res.status(200).send({
+          msg: 'Your image was uploaded sucessfully',
+          fileName: req.file.key,
+        });
+        const AwsUrl = `https://${BUCKET}.s3.amazonaws.com/${req.file.key}`
+        db.User.updateOne({ _id: req.userId }, { $set: { photoUrl: AwsUrl } })
+          .then((response) => {
+            console.log(response)
+            res.send(response) 
+          })
+          .catch((error) => {
+            console.log(error);
+            res.status(500).send(error);
+          });
+      }
```
When I ran the app, I'm getting an `Cannot set headers after they are sent to the client` error. I think these few lines might be why because you are attempting to send the response in `.then()` and `.catch()` after sending the response already. 

I suggest putting line 39-42 inside the `then` block. That way, you can guarantee that the url has been stored in the database before telling user that image is uploaded.

mattchx committed:
Merge branch 'dev' into feature/photo-upload
